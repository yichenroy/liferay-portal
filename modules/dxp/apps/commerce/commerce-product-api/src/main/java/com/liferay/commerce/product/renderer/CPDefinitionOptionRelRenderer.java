/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 * <p>
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * <p>
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.renderer;

import aQute.bnd.annotation.ProviderType;

import com.liferay.portal.kernel.exception.PortalException;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * @author Marco Leo
 */
@ProviderType
public interface CPDefinitionOptionRelRenderer {

	public String render(
		long cpDefinitionId, HttpServletRequest httpServletRequest,
		HttpServletResponse httpServletResponse) throws PortalException;

}