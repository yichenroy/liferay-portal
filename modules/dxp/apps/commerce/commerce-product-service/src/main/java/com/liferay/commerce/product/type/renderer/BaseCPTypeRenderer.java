/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.type.renderer;

import com.liferay.commerce.product.type.CPTypeRenderer;
import com.liferay.portal.kernel.log.Log;
import com.liferay.portal.kernel.log.LogFactoryUtil;
import com.liferay.portal.kernel.model.PortletConstants;
import com.liferay.portal.kernel.portlet.PortletBag;
import com.liferay.portal.kernel.portlet.PortletBagPool;
import com.liferay.portal.kernel.util.Portal;
import com.liferay.portal.kernel.util.Validator;
import com.liferay.portal.kernel.util.WebKeys;

import java.io.IOException;

import javax.servlet.RequestDispatcher;
import javax.servlet.ServletContext;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.osgi.service.component.annotations.Reference;

/**
 * @author Marco Leo
 */
public abstract class BaseCPTypeRenderer implements CPTypeRenderer {

	public void renderJSP(
			HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse, String path)
		throws IOException {

		renderJSP(
			getServletContext(httpServletRequest), httpServletRequest,
			httpServletResponse, path);
	}

	public void renderJSP(
			ServletContext servletContext,
			HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse, String path)
		throws IOException {

		RequestDispatcher requestDispatcher =
			servletContext.getRequestDispatcher(path);

		try {
			requestDispatcher.include(httpServletRequest, httpServletResponse);
		}
		catch (ServletException se) {
			_log.error("Unable to render JSP " + path, se);

			throw new IOException("Unable to render " + path, se);
		}
	}

	public void setServletContext(ServletContext servletContext) {
		_servletContext = servletContext;
	}

	protected ServletContext getServletContext(
		HttpServletRequest httpServletRequest) {

		if (_servletContext != null) {
			return _servletContext;
		}

		String portletId = portal.getPortletId(httpServletRequest);

		if (Validator.isNotNull(portletId)) {
			String rootPortletId = PortletConstants.getRootPortletId(portletId);

			PortletBag portletBag = PortletBagPool.get(rootPortletId);

			return portletBag.getServletContext();
		}

		return (ServletContext)httpServletRequest.getAttribute(WebKeys.CTX);
	}

	@Reference
	protected Portal portal;

	private static final Log _log = LogFactoryUtil.getLog(
		BaseCPTypeRenderer.class);

	private ServletContext _servletContext;

}