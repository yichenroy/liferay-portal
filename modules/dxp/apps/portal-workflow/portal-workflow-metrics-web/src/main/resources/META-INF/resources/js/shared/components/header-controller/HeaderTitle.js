import PortalComponent from './PortalComponent';
import React from 'react';

export default class HeaderTitle extends React.Component {
	componentDidUpdate({title: prevTitle}) {
		const {title} = this.props;

		if (prevTitle != title) {
			this.setDocumentTitle(prevTitle, title);
		}
	}

	setDocumentTitle(prevTitle, title) {
		document.title = document.title.replace(prevTitle, title);
	}

	render() {
		const {container, title} = this.props;

		return (
			<PortalComponent container={container} replace>
				{title}
			</PortalComponent>
		);
	}
}