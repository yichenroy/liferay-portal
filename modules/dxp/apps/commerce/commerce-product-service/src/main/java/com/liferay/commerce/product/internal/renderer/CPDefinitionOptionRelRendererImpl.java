/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.commerce.product.internal.renderer;

import com.liferay.commerce.product.model.CPDefinitionOptionRel;
import com.liferay.commerce.product.model.CPDefinitionOptionValueRel;
import com.liferay.commerce.product.renderer.CPDefinitionOptionRelRenderer;
import com.liferay.commerce.product.service.CPDefinitionOptionRelService;
import com.liferay.dynamic.data.mapping.form.renderer.DDMFormRenderer;
import com.liferay.dynamic.data.mapping.form.renderer.DDMFormRenderingContext;
import com.liferay.dynamic.data.mapping.model.DDMForm;
import com.liferay.dynamic.data.mapping.model.DDMFormField;
import com.liferay.dynamic.data.mapping.model.DDMFormFieldOptions;
import com.liferay.dynamic.data.mapping.model.LocalizedValue;
import com.liferay.portal.kernel.exception.PortalException;
import com.liferay.portal.kernel.util.Portal;

import java.util.List;
import java.util.Locale;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.osgi.service.component.annotations.Component;
import org.osgi.service.component.annotations.Reference;

/**
 * @author Marco Leo
 */
@Component(immediate = true)
public class CPDefinitionOptionRelRendererImpl
	implements CPDefinitionOptionRelRenderer {

	@Override
	public String render(
			long cpDefinitionId, HttpServletRequest httpServletRequest,
			HttpServletResponse httpServletResponse)
		throws PortalException {

		DDMForm ddmForm = new DDMForm();

		Locale locale = _portal.getLocale(httpServletRequest);

		List<CPDefinitionOptionRel> cpDefinitionOptionRels =
			_cpDefinitionOptionRelService.
				getSkuContributorCPDefinitionOptionRels(cpDefinitionId);

		for (CPDefinitionOptionRel cpDefinitionOptionRel :
				cpDefinitionOptionRels) {

			List<CPDefinitionOptionValueRel> cpDefinitionOptionValueRels =
				cpDefinitionOptionRel.getCPDefinitionOptionValueRels();

			DDMFormField ddmFormField = new DDMFormField(
				String.valueOf(cpDefinitionOptionRel.
					getCPDefinitionOptionRelId()),
				cpDefinitionOptionRel.getDDMFormFieldTypeName());

			if (!cpDefinitionOptionValueRels.isEmpty()) {
				DDMFormFieldOptions ddmFormFieldOptions =
					new DDMFormFieldOptions();

				for (CPDefinitionOptionValueRel cpDefinitionOptionValueRel :
						cpDefinitionOptionValueRels) {

					ddmFormFieldOptions.addOptionLabel(
						String.valueOf(cpDefinitionOptionValueRel.
							getCPDefinitionOptionValueRelId()),
						locale, cpDefinitionOptionValueRel.getTitle(locale));
				}

				ddmFormField.setDDMFormFieldOptions(ddmFormFieldOptions);
			}

			LocalizedValue localizedValue = new LocalizedValue(locale);

			localizedValue.addString(
				locale, cpDefinitionOptionRel.getName(locale));

			ddmFormField.setLabel(localizedValue);

			ddmFormField.setRequired(true);

			ddmForm.addDDMFormField(ddmFormField);
		}

		DDMFormRenderingContext ddmFormRenderingContext =
			new DDMFormRenderingContext();

		ddmFormRenderingContext.setContainerId("options");
		ddmFormRenderingContext.setHttpServletRequest(httpServletRequest);
		ddmFormRenderingContext.setHttpServletResponse(httpServletResponse);
		ddmFormRenderingContext.setLocale(locale);

		return _ddmFormRenderer.render(ddmForm, ddmFormRenderingContext);
	}

	@Reference
	private CPDefinitionOptionRelService _cpDefinitionOptionRelService;

	@Reference
	private DDMFormRenderer _ddmFormRenderer;

	@Reference
	private Portal _portal;

}