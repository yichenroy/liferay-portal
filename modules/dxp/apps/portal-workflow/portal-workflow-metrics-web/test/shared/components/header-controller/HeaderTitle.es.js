/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Liferay Enterprise
 * Subscription License ("License"). You may not use this file except in
 * compliance with the License. You can obtain a copy of the License by
 * contacting Liferay, Inc. See the License for the specific language governing
 * permissions and limitations under the License, including but not limited to
 * distribution rights of the Software.
 */

import HeaderTitle from '../../../../src/main/resources/META-INF/resources/js/shared/components/header-controller/HeaderTitle.es';
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

beforeAll(() => {
	const vbody = document.createElement('div');

	vbody.innerHTML = '<div id="workflow"></div>';
	document.body.appendChild(vbody);

	ReactDOM.createPortal = jest.fn(element => {
		return element;
	});
});

test('Should render component on container', () => {
	const container = document.getElementById('workflow');

	const component = renderer.create(
		<HeaderTitle container={container} title="Metrics" />
	);

	const tree = component.toJSON();

	expect(tree).toMatchSnapshot();
});

test('Should set document title', () => {
	document.title = 'Metrics';

	const container = document.getElementById('workflow');

	const component = shallow(
		<HeaderTitle container={container} title="Metrics" />
	);

	const instance = component.instance();

	instance.setDocumentTitle('Metrics', 'SLAs');

	expect(document.title).toEqual('SLAs');
});

test('Should set document title if title prop changed', () => {
	document.title = 'Metrics';

	const container = document.getElementById('workflow');

	const component = shallow(
		<HeaderTitle container={container} title="Metrics" />
	);

	component.setProps({title: 'SLAs'});

	expect(document.title).toEqual('SLAs');
});
