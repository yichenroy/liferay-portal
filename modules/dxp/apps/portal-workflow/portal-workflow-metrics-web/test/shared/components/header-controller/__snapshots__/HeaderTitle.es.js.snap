// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render component on container 1`] = `"Metrics"`;
